core/src/main/java/org/pentaho/di/core/Props.java

```diff
@@ -0,0 +1,672 @@
+  }
+
+  protected Props() {
+    init();
+  }
+
+  protected Props( int t ) {
+    type = t;
+    filename = getFilename();
+    init();
+  }
+
+  protected void init() {
+    createLogChannel();
+    properties = new Properties();
+    pluginHistory = new ArrayList<ObjectUsageCount>();
+
+    loadProps();
+    addDefaultEntries();
+
+    loadPluginHistory();
+
+  }
+
+  protected Props( String filename ) {
+    properties = new Properties();
+    this.type = TYPE_PROPERTIES_EMPTY;
+    this.filename = filename;
+    init();
+  }
+
+  @Override
+  public String toString() {
+    return STRING_USER_PREFERENCES;
+  }
+
+  protected void createLogChannel() {
+    log = new LogChannel( STRING_USER_PREFERENCES );
+  }
+
+  public String getFilename() {
+    String filename = "";
+    String directory = Const.getKettleDirectory();
+
+    switch ( type ) {
+      case TYPE_PROPERTIES_SPOON:
+      case TYPE_PROPERTIES_PAN:
+        filename = directory + Const.FILE_SEPARATOR + ".spoonrc";
+        break;
+      case TYPE_PROPERTIES_CHEF:
+      case TYPE_PROPERTIES_KITCHEN:
+        filename = directory + Const.FILE_SEPARATOR + ".chefrc";
+        break;
+      case TYPE_PROPERTIES_MENU:
+        filename = directory + Const.FILE_SEPARATOR + ".menurc";
+        break;
+      case TYPE_PROPERTIES_PLATE:
+        filename = directory + Const.FILE_SEPARATOR + ".platerc";
+        break;
+      default:
+        break;
+    }
+
+    return filename;
+  }
+
+  public String getLicenseFilename() {
+    String directory = Const.getKettleDirectory();
+    String filename = directory + Const.FILE_SEPARATOR + ".licence";
+
+    // Try to create the directory...
+    File dir = new File( directory );
+    if ( !dir.exists() ) {
+      try {
+        dir.mkdirs();
+      } catch ( Exception e ) {
+        // ignore - should likely report failure to create directory
+      }
+    }
+
+    return filename;
+  }
+
+  public boolean fileExists() {
+    File f = new File( filename );
+    return f.exists();
+  }
+
+  public void setType( int t ) {
+    type = t;
+  }
+
+  public int getType() {
+    return type;
+  }
+
+  public boolean loadProps() {
+    try {
+      FileInputStream fis = new FileInputStream( filename );
+      try {
+        properties.load( fis );
+      } finally {
+        try {
+          fis.close();
+        } catch ( IOException ignored ) {
+          // Ignore close exception
+        }
+      }
+    } catch ( Exception e ) {
+      return false;
+    }
+    return true;
+  }
+
+  protected void addDefaultEntries() {
+    if ( !properties.containsKey( "JobDialogStyle" ) ) {
+      properties.setProperty( "JobDialogStyle", "RESIZE,MAX,MIN" );
+    }
+  }
+
+  public void saveProps() {
+
+    File spoonRc = new File( filename );
+    try {
+      // FileOutputStream fos = new FileOutputStream(spoonRc);
+
+      SortedFileOutputStream fos = new SortedFileOutputStream( spoonRc );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code contains a resource that might not have closed properly. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *fos*. The resource is referenced at lines 351, 352. The resource is closed at line 353. There are other execution paths that do not contain closure statements, for example, when Properties.store throws an exception.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).